pkg/controllers/provisioning/provisioner.go

```diff
@@ -89,28 +71,44 @@ type Provisioner struct {
 }
 
 // Add a pod to the provisioner and block until it's processed. The caller
-// is responsible for verifying that the pod was scheduled correctly. In the
-// future, this may be expanded to include concepts such as retriable errors.
-func (p *Provisioner) Add(ctx context.Context, pod *v1.Pod) {
-	select {
-	case p.pods <- pod: // Block until pod is enqueued
-		<-p.wait
-	case <-p.running.Done(): // Leave if closed
+// is responsible for verifying that the pod was scheduled correctly.
+func (p *Provisioner) Add(ctx context.Context, pod *v1.Pod) <-chan struct{} {
+	p.RLock()
+	defer p.RUnlock()
+	p.batch.Add(pod)
+	return p.batch.Done
+}
+
+func (p *Provisioner) run(ctx context.Context) {
+	if ctx.Err() != nil {
+		logging.FromContext(ctx).Info("Stopping provisioner")
+		return
+	}
+	if err := p.provision(ctx); err != nil {
+		logging.FromContext(ctx).Errorf("Provisioning failed, %s", err.Error())
 	}
+	p.run(ctx)
 }
 
 func (p *Provisioner) provision(ctx context.Context) (err error) {
-	// Wait for a batch of pods, release when done
-	pods := p.batch(ctx)
-	defer func() {
-		for i := 0; i < len(pods); i++ {
-			p.wait <- struct{}{}
+	// Batch pods
+	logging.FromContext(ctx).Infof("Waiting for unschedulable pods")
+	items, window := p.batch.Wait(ctx)
+	defer p.batch.Complete()
+	p.Lock()
+	p.batch = NewBatch(ctx)
```

Comment:
I don't quite get why the locks are here?  

It seems like this would wait until nothing is being Added, and then create a new batch, but something could have been added before the Lock was acquired.    Wouldn't it be easier to control this if you used the same batch so that it could be continuous rather than having to move between batches?

Reply:
This is an interesting idea. I can move the locking mechanism inside the batch and reuse it. I have mutation either way. Let me play with this.